[CORL-2222/2223/2225] Add story archiving features to Coral

nick-funk opened the pull request:
## What does this PR do?

Adds the ability to archive stories to a secondary mongo instance.

Adds the mutations to archive stories and also adds the auto-archiving job features to let stories archive as they age and pass a certain age.

## What changes to the GraphQL/Database Schema does this PR introduce?

Adds two mutations to `archiveStories` and `unarchiveStories` along with their input's and payloads.

## How do I test this PR?

- Spin up another mongo instance to be your archiving Db
```
docker run -d -p 27018:27017 --restart always --name mongo-archive mongo:4.2
```
- Set the environment variable for this new archiving URI
```
MONGODB_ARCHIVE_URI=mongodb://127.0.0.1:27018/coral
```
- Find a story to archive and use the following mutation to do so:
```
mutation($storyIDs: [ID!]!, $clientMutationId: String!) {
  archiveStories(input: {
    storyIDs: $storyIDs,
    clientMutationId: $clientMutationId
  }) {
    stories {
      id
      isArchiving
      isArchived
    }
  }
}

// Query variables
{
	"storyIDs": [
		"<STORY_ID_YOU_WANT_TO_ARCHIVE>"
	],
	"clientMutationId": "0"
}
```
- Go check the story in stream and see that it is closed, archived, and is completely read-only
 
## How do we deploy this PR?

- If a client is not using archiving, you don't need to do anything.

- If you want to only use archiving and only use manual archive mutations on a client, you need to set the archive URI env variable on their regular and job deploy yaml:

```
MONGODB_ARCHIVE_URI
```
(if this is not set, it will default to the main mongodb URI.)

- If you want to use auto-archiving job features, you need to set the following on the job deploy yaml (or whichever instance will do archiving jobs).

```
ENABLE_AUTO_ARCHIVING <-- set to true to turn archive jobs on
AUTO_ARCHIVE_OLDER_THAN <-- optional, has a default
AUTO_ARCHIVING_BATCH_SIZE <-- optional, has a default
```

nick-funk committed:
Create initial archiving connections, mutations

nick-funk committed:
Add in second archive mongo connection

nick-funk committed:
Hook up archiving mutators and some initial resolvers

nick-funk committed:
Update comment counts when archiving/un-archiving

nick-funk committed:
Preliminarily update stream resolvers to use archives

nick-funk committed:
Initial pass at disabling stream actions on archived stories

nick-funk committed:
Make single conversation view work with archived comments

nick-funk committed:
Update delete user to work with archived comments

nick-funk committed:
Update tests

nick-funk committed:
Refactor to aggregate archive count updates with a multiplier

nick-funk committed:
Create an auto-archiving task that will run once a day

nick-funk committed:
Tighten up safety locks around archiving/un-archiving

nick-funk committed:
Rename AUTO_ARCHIVING_AMOUNT to AUTO_ARCHIVING_BATCH_SIZE

nick-funk committed:
Create data context to store mongo, archive, and redis connections

nick-funk committed:
Create new shared context for the two Mongo clients to pass through

nick-funk committed:
Use for of instead of .forEach when auto-archiving stories

This makes the operation more trace-able in both logs and
in repeating the auto-archive batch if an error occurs and
we need to debug.

nick-funk committed:
Fix issue where archive job could end abruptly

Previous version of batch iteration was a .forEach
whereas now it is a for of loop. We can't return out
of a for..of without breaking the batch loop.

Instead switch to checking on lock prior to archiving
so we don't need a return/continue and add comments
to clarify how this works.

nick-funk committed:
Improve searching for auto-archive stories

Uses same lock improvements from prior commit adding
improvements to the markStoryForArchive.

nick-funk committed:
Allow archiving multiple stories with archive mutations

wyattjoh commented on src/core/client/stream/tabs/Comments/Comment/CommentContainer.tsx:
```diff
@@ -643,7 +653,10 @@ export const CommentContainer: FunctionComponent<Props> = ({
-                          settings.disableCommenting.enabled || story.isClosed
+                          settings.disableCommenting.enabled ||
+                          story.isClosed ||
+                          story.isArchived ||
+                          story.isArchiving
```
Aren't all stories that are archived (or archiving) also closed?

nick-funk replied:
Yes

wyattjoh commented on src/core/client/stream/tabs/Configure/OpenOrCloseStream/OpenOrCloseStreamContainer.tsx:
```diff
@@ -48,6 +48,10 @@ const OpenOrCloseStreamContainer: FunctionComponent<Props> = ({
+  if (story.isArchiving || story.isArchived) {
+    return null;
+  }
+
```
Maybe instead it would be best to show a readonly button with a warning that "this story is archived"? Clarification from @losowsky obviously.

nick-funk replied:
Yeah, we're still waiting for Andrew to get back. This was brought up by myself and others in the "what status do we want to show for archiving?" and then we realized both Sam and Andrew were out, so we're sitting on this for now.

wyattjoh commented on src/core/server/cron/accountDeletion.ts:
```diff
@@ -14,6 +14,7 @@ import {
 
 interface Options {
   mongo: Db;
+  archive: Db;
```
Why not take the combined Mongo context here?

wyattjoh commented on src/core/server/data/context.ts:
```diff
@@ -0,0 +1,28 @@
+import { AugmentedRedis } from "coral-server/services/redis";
+
+export interface MongoContext {
+  readonly main: Db;
```
I almost like the `live` used in the options better than `main`? `live` is more meaningful.

nick-funk replied:
I originally called it `live` and @cvle asked me to change it to `main` to avoid confusion. You two can decide what you prefer.

wyattjoh commented on src/core/server/data/context.ts:
```diff
@@ -0,0 +1,28 @@
+
+interface DataContextOptions {
+  mongoLive: Db;
+  mongoArchive: Db;
```
You could possibly embed this a bit further to simplify the access:

```suggestion
  mongo: MongoContext;
```

nick-funk replied:
Alright, this and your other comment makes me feel like I should go all in and refactor this context through the whole code base. No worries, I hear you and it will be done.

wyattjoh commented on src/core/server/graph/loaders/Comments.ts:
```diff
@@ -190,21 +191,34 @@ export default (ctx: GraphContext) => ({
+  }: QueryToCommentsArgs) => {
+    let story: Readonly<Story> | null = null;
+    if (storyID) {
+      story = await retrieveStory(ctx.mongo, ctx.tenant.id, storyID);
```
Should instead use:

```suggestion
      story = await ctx.loaders.Stories.story.load(storyID);
```

wyattjoh commented on src/core/server/graph/mutators/Stories.ts:
```diff
@@ -140,4 +149,56 @@ export const Stories = (ctx: GraphContext) => ({
+      stories.push(result);
+    }
+
+    return stories;
```
This logic should live in a new `archiving` service folder.

wyattjoh commented on src/core/server/graph/mutators/Stories.ts:
```diff
@@ -140,4 +149,56 @@ export const Stories = (ctx: GraphContext) => ({
+      stories.push(result);
+    }
+
+    return stories;
```
Same here, should live in a new `archiving` service folder.

wyattjoh commented on src/core/server/data/context.ts:
```diff
@@ -0,0 +1,28 @@
+  redis: AugmentedRedis;
+}
+
+export default class DataContext {
```
It's common GraphQL lingo to refer to this as a `Connector`, so this should be called `Connectors`.

wyattjoh commented on src/core/server/graph/context.ts:
```diff
@@ -86,6 +89,8 @@ export default class GraphContext {
   public readonly signingConfig?: JWTSigningConfig;
   public readonly user?: User;
 
+  public readonly dataContext: DataContext;
```
Following the feedback around renaming the `DataContext` class to `Connectors`, similarrly this would then become:

```suggestion
  public readonly connectors: Connectors;
```

wyattjoh commented on src/core/server/graph/context.ts:
```diff
@@ -104,6 +109,7 @@ export default class GraphContext {
     this.i18n = options.i18n;
     this.pubsub = options.pubsub;
     this.mongo = options.mongo;
+    this.archive = options.archive;
```
If the connectors are being added to the context directly, you should remove the reference to these specifically.

wyattjoh commented on src/core/server/index.ts:
```diff
@@ -221,7 +224,19 @@ class Server {
+    }
+
+    this.mongo = await createMongoDB(mongoURI);
+    this.archive = await createMongoDB(archiveMongoURI);
```
I would rather see this be a bit different! Right now, if the same URL is for both the archive and the mongo instance, we actually create 2X the connections to the same instance! Instead, I'd suggest the following logic:

```suggestion
    this.mongo = await createMongoDB(mongoURI);

    const mongoURI = config.get("mongodb");
    let archiveMongoURI = config.get("mongodbArchive");
    // if the archive URI isn't specified, but the regular mongo uri
    // is, then use the same db for both mongo instances
    if (
      archiveMongoURI === config.default("mongodbArchive") &&
      mongoURI !== config.default("mongodb")
    ) {
      this.archive = this.mongo;
    } else {
      this.archive = await createMongoDB(archiveMongoURI);
    }
```

wyattjoh commented on src/core/server/models/comment/comment.ts:
```diff
@@ -467,7 +511,7 @@ export const retrieveCommentRepliesConnection = (
  */
 export async function retrieveCommentParentsConnection(
-  mongo: Db,
+  mongo: MongoContext,
```
I think for this function, it could additionally get a `isArchived: boolean` to help pass down to the `retrieveManyComments`.

wyattjoh commented on src/core/server/models/comment/comment.ts:
```diff
@@ -388,24 +396,54 @@ export async function editComment(
 
 export async function retrieveManyComments(
-  mongo: Db,
+  mongo: MongoContext,
```
This should instead have a `isArchived` parameter to avoid the double pulling for this scenario.

wyattjoh commented on src/core/server/models/comment/comment.ts:
```diff
@@ -388,24 +396,54 @@ export async function editComment(
 
-  const comments = await cursor.toArray();
+  const liveComments = await liveCommentsCursor.toArray();
+  if (liveComments && liveComments.length > 0) {
```
```suggestion
  if (liveComments.length > 0) {
```

wyattjoh commented on src/core/server/models/comment/comment.ts:
```diff
@@ -388,24 +396,54 @@ export async function editComment(
   ids: ReadonlyArray<string>
 ) {
-  const cursor = collection(mongo).find({
+  // Try and find it in live comments collection
```
Best to add a guard at the top to prevent unneeded database calls:

```ts
if (ids.length === 0) {
  return []
}
```

wyattjoh commented on src/core/server/models/story/story.ts:
```diff
@@ -727,3 +742,306 @@ export async function retrieveStorySections(
   // the null entries, the resulting array can not contain null.
   return results.filter((section) => section !== null).sort() as string[];
 }
+
```
I would strongly suggest extracting this to a service function, and exposing the MongoDB methods as methods that are imported from each model module. Reaching between collections in a single model file should be avoided at all costs.

wyattjoh commented on src/core/server/models/story/story.ts:
```diff
@@ -727,3 +742,306 @@ export async function retrieveStorySections(
+      tenantID,
+      $and: [
+        { $or: [{ isArchiving: { $exists: false } }, { isArchiving: false }] },
+        { $or: [{ isArchived: { $exists: false } }, { isArchived: false }] },
```
This can be more compactly written:

```suggestion
        { isArchiving: { $in: [null, false] } },
        { isArchived: { $in: [null, false] } },
```

nick-funk replied:
Thanks, will update!

wyattjoh commented on src/core/server/models/story/story.ts:
```diff
@@ -727,3 +742,306 @@ export async function retrieveStorySections(
+      if (result.action.hasOwnProperty.call(commentCounts.action, key)) {
+        const value = commentCounts.action[key];
+        result.action[key] = value * multiplier;
+      }
```
Best to invert this for more clarity:

```suggestion
      if (!Object.prototype.hasOwnProperty.call(commentCounts.action, key)) {
        continue;
      }
      
      const value = commentCounts.action[key];
      result.action[key] = value * multiplier;
```

wyattjoh replied:
Apply the same changes below!

nick-funk replied:
👍

wyattjoh commented on src/core/server/models/story/story.ts:
```diff
@@ -727,3 +742,306 @@ export async function retrieveStorySections(
+  return result.value;
+}
+
+const getCommentCounts = (options: {
```
Best to move this into the `src/core/server/models/comment/counts/counts.ts` file as well.. Call it something like `negateCommentCounts`.

nick-funk replied:
💯 👍

wyattjoh commented on src/core/server/models/story/story.ts:
```diff
@@ -727,3 +742,306 @@ export async function retrieveStorySections(
+    throw new StoryNotFoundError(id);
+  }
+
+  if (targetStory.isArchived) {
```
Maybe also check if the story is currently already being archived?

nick-funk replied:
It will always be archiving cause of the way the `markStoryForArchiving` and `markStoryForUnarchiving` locks the story.

wyattjoh commented on src/core/server/models/story/story.ts:
```diff
@@ -727,3 +742,306 @@ export async function retrieveStorySections(
+    await archivedCommentModerationActions(mongo.archive).insertMany(
+      targetCommentModerationActions
+    );
+  }
```
I have a bit of an issue with how _all_ this data is loaded into memory and only then dumped out to each collection. I'd rather instead see something where you could batch these into more manageable chunks. This would be why the "comments" model could instead do something like:

```
cursor = < create cursor of comments to archive >
batch = []
for comment in cursor:
  batch.push(comment)
  if batch.length >= BATCH_MAX:
    archivedComments.insertMany(batch);
    batch = []
if batch.length > 0:
  archivedComments.insertMany(batch);
  batch = []
```

That way the documents are handed over in reasonable batches. `toArray` does this batching internally, but it just keeps going until the cursor is consumed.

nick-funk replied:
This has been batched

nick-funk committed:
Move the archiving operations to a service

nick-funk committed:
Use a batched document move for archiving tasks

nick-funk committed:
Add further logging to the archive/unarchive operations

nick-funk committed:
Use bulk ops and tenant filter during document removals

nick-funk committed:
Use unordered bulk ops to insert during un/archive tasks

nick-funk committed:
Fix returning nulls in archive stories graphql response

nick-funk committed:
Update App and Graph context to use MongoContext

No longer two distinct Db's of mongo and archive, just
one MongoContext called mongo with both live and archive
inside of it.

It is then passed down accordingly throughout the rest of
the application.

This is giant, but mostly benign "touching of every file" in
the back-end.

nick-funk committed:
Rename MongoContext.main to live

nick-funk committed:
Only use 1 mongo connection if archiving isn't set up

nick-funk committed:
Don't allow auto archiving to re-archive un-archived stories

nick-funk committed:
Show archived state in admin > stories

nick-funk committed:
Allow archive/unarchive from the admin > stories page

nick-funk committed:
Show zeroed out mod queues when filtering by an archived story

nick-funk committed:
Show alternate stream configure flow for archived stories

nick-funk committed:
Show archived pill instead of "moderate" on archived streams

nick-funk committed:
Show different report form for archived stories

nick-funk committed:
Set fixture story url to fix some failing tests

nick-funk committed:
Allow users to determine how often auto archiving is run

Is set through an env variable AUTO_ARCHIVING_INTERVAL whose
value is a cron timing.

nick-funk committed:
Update create/reply mutations to return story url in optimistic response

nick-funk committed:
Improving error logging for test console alerts

nick-funk committed:
Don't allow opening a story if it's archived or archiving

nick-funk committed:
Translate archived story reported emails

wyattjoh commented on src/core/server/app/middleware/passport/strategies/jwt.ts:
```diff
@@ -60,8 +60,8 @@ export function createVerifiers(
-    new SSOVerifier(options),
-    new JWTVerifier(options),
+    new SSOVerifier({ ...options, mongo: options.mongo.live }),
+    new JWTVerifier({ ...options, mongo: options.mongo.live }),
```
Looks like you adjusted the `OIDCVerifier` but not these two.

wyattjoh commented on src/core/server/graph/loaders/Comments.ts:
```diff
@@ -190,21 +190,34 @@ export default (ctx: GraphContext) => ({
+  }: QueryToCommentsArgs) => {
+    let story: Readonly<Story> | null = null;
+    if (storyID) {
+      story = await retrieveStory(ctx.mongo.live, ctx.tenant.id, storyID);
```
Should use the loader here:

```suggestion
      story = await ctx.loaders.Stories.story.load(storyID)
```

wyattjoh commented on src/core/server/graph/mutators/Stories.ts:
```diff
@@ -138,6 +151,77 @@ export const Stories = (ctx: GraphContext) => ({
+      if (result) {
+        stories.push(result);
+      }
+    }
```
This logic should be moved to "coral-server/services/archive".

wyattjoh commented on src/core/server/models/comment/comment.ts:
```diff
@@ -353,7 +357,11 @@ export async function editComment(
+      { live: mongo, archive: mongo },
+      tenantID,
+      id
+    );
```
In this case, I'd suggest either making a separate function for `retrieveComment` that only accepts a single `mongo: Db`, as realistically, this is adding an extra DB call for an edge case. Then you could modify it to instead look like:

```ts
function isMongoContext(mongo: Db | MongoContext): mongo is MongoContext {
  return (
    typeof (mongo as MongoContext).live !== "undefined" &&
    (mongo as MongoContext).live instanceof Db &&
    typeof (mongo as MongoContext).archive !== "undefined" &&
    (mongo as MongoContext).archive instanceof Db
  );
}

export async function retrieveComment(
  mongo: MongoContext | Db,
  tenantID: string,
  id: string
) {
  const live = isMongoContext(mongo) ? mongo.live : mongo;
  const archive = isMongoContext(mongo) ? mongo.archive : null;

  const liveComment = await collection(live).findOne({ id, tenantID });
  if (liveComment) {
    return liveComment;
  }

  // If the passed mongo object wasn't a context, we have no archive to check!
  if (!archive) {
    return null;
  }

  const archivedComment = await archivedComments(archive).findOne({
    id,
    tenantID,
  });
  return archivedComment;
}
```

So you end up writing:

```suggestion
    const comment = await retrieveComment(mongo, tenantID, id);
```

Just an idea! I just don't like the idea of always checking the archive (for this use case) which for editing a comment, can't happen anyways.

wyattjoh replied:
But actually move `MongoContext` use to _only_ services.

wyattjoh commented on src/core/server/cron/autoArchiving.ts:
```diff
@@ -0,0 +1,109 @@
+    return;
+  }
+
+  const amount = config.get("auto_archiving_amount");
```
`BATCH_SIZE` is more correct.

wyattjoh commented on src/core/server/cron/autoArchiving.ts:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  const amount = config.get("auto_archiving_amount");
+  const age = config.get("auto_archiving_age");
```
`auto_archive_older_than`

wyattjoh commented on src/core/server/index.ts:
```diff
@@ -221,14 +222,30 @@ class Server {
+      archiveURI === config.default("mongodb_archive") &&
+      liveURI !== config.default("mongodb")
+    ) {
+      archive = live;
```
Upon further reflection, I'd argue that an archive feature that uses the same db connection isn't helpful. While it's useful in development to archive to the same db instance, it provides none of the expected performance benefits.

wyattjoh commented on src/core/server/models/story/helpers.ts:
```diff
@@ -67,6 +67,20 @@ export function getStoryClosedAt(
+}
+
+export function isStoryArchived(story: Pick<Story, "isArchived">) {
+  return story.isArchived || false;
```
```suggestion
  return story.isArchived ?? false;
```

wyattjoh commented on src/core/server/cron/autoArchiving.ts:
```diff
@@ -0,0 +1,109 @@
+
+    const now = new Date();
+    const dateFilter = new Date(now.getTime() - age);
+    const stories = await retrieveStoriesToBeArchived(
```
Risk! By bulk fetching stories and marking them at once means if a failure occurs, you might be in a larger broken state than one at a time!

nick-funk replied:
They're not being marked, but they are being queried out as a list. I'll make this walk a cursor instead.

wyattjoh commented on src/core/server/cron/autoArchiving.ts:
```diff
@@ -0,0 +1,109 @@
+      mongo.live,
+      tenant.id,
+      dateFilter,
+      amount
```
```ts
for (let j = 0; j < amount; j++) {
```

And get one story at a time.

wyattjoh commented on src/core/server/models/story/story.ts:
```diff
@@ -727,3 +735,81 @@ export async function retrieveStorySections(
+        isArchiving: true,
+        closedAt: now,
+        updatedAt: now,
+        archivedAt: now,
```
`startedArchivingAt` instead! Should add a `TODO` around handling partially archived stories.

wyattjoh commented on src/core/server/services/archive/index.ts:
```diff
@@ -0,0 +1,328 @@
+const getCommentCounts = (options: {
+  story: Readonly<Story>;
+  negate: boolean;
+}) => {
```
```suggestion
const negateCommentCounts = (commentCounts: Readonly<RelatedCommentCounts>) => {
```

wyattjoh replied:
Also, move to `src/core/server/models/comment/counts/counts.ts` please!

wyattjoh commented on src/core/server/services/archive/index.ts:
```diff
@@ -0,0 +1,328 @@
+
+  if (commentCounts.action) {
+    for (const key in commentCounts.action) {
+      if (result.action.hasOwnProperty.call(commentCounts.action, key)) {
```
```suggestion
      if (Object.prototype.hasOwnProperty.call(commentCounts.action, key)) {
```

For the rest of this file please!

wyattjoh commented on src/core/server/services/archive/index.ts:
```diff
@@ -0,0 +1,328 @@
+  const targetCommentIDs = await moveDocuments({
+    tenantID,
+    source: comments(mongo.live),
+    selectionCursor: targetComments,
```
Can use https://github.com/coralproject/talk/blob/develop/src/core/server/models/helpers/query.ts#L12 instead!

wyattjoh commented on src/core/server/services/archive/index.ts:
```diff
@@ -0,0 +1,328 @@
+    {
+      returnOriginal: false,
+    }
+  );
```
Move to the story model!

wyattjoh commented on src/core/server/services/archive/index.ts:
```diff
@@ -0,0 +1,328 @@
+    {
+      returnOriginal: false,
+    }
+  );
```
Move to models!

wyattjoh commented on src/core/server/services/archive/index.ts:
```diff
@@ -0,0 +1,328 @@
+  selectionCursor: Cursor;
+  destination: Collection;
+  returnMovedIDs?: boolean;
+}
```
```suggestion
interface MoveDocumentsOptions<T> {
  tenantID: string;
  source: Collection<T>;
  fulter: FilterQuery<T>;
  destination: Collection<T>;
  returnMovedIDs?: boolean;
}
```

wyattjoh commented on src/core/server/services/archive/index.ts:
```diff
@@ -0,0 +1,328 @@
+  destination,
+  returnMovedIDs = false,
+}: MoveDocumentsOptions) => {
+  let insertBatch: any[] = [];
```
```suggestion
async function moveDocuments<T extends { id: string }>({
  tenantID,
  source,
  filter,
  destination,
  returnMovedIDs = false,
}: MoveDocumentsOptions<T>) {
  let insertBatch: T[] = [];
```

wyattjoh commented on src/core/server/services/archive/index.ts:
```diff
@@ -0,0 +1,328 @@
+
+    insertBatch.push(document);
+
+    const hasID = Object.prototype.hasOwnProperty.call(document, "id");
```
Not required now!

wyattjoh commented on src/core/server/index.ts:
```diff
@@ -221,14 +222,30 @@ class Server {
+      archive = await createMongoDB(archiveURI);
+    }
+
+    this.mongo = { live, archive };
```
Move the creation logic inside the context, so you'll have a `createMongoContext(config)`

nick-funk committed:
Standardize the dependency injection for verifiers

nick-funk committed:
Create service and model levels for comment retrieval

Separates the concerns of the comments retrieval of
archived vs live out of the model layer and into the
service layer for comments.

nick-funk committed:
Rename and move getCommentCounts to negateCommentCounts

nick-funk committed:
Use generics to simplify moveDocuments method

nick-funk committed:
Create markStoryAsArchived, markStoryAsUnarchived functions

nick-funk committed:
Better date records when archiving

nick-funk committed:
Fix single conversation view throwing data loader error

nick-funk committed:
Retrieve archived parent comments in single convo view

nick-funk committed:
Merge branch 'develop' into spike/archiving

nick-funk committed:
Rename auto archive env variables for clarity

nick-funk committed:
Use null coalescing operator for nullable

nick-funk committed:
Make autoArchive walk through a story cursor

Allows us to find and mark one story at a time
for better repeatability of the auto-archiving
process.

alinhle commented on src/core/client/admin/routes/Stories/StoryActions/ArchiveStoriesMutation.ts:
```diff
@@ -0,0 +1,54 @@
+      },
+      variables: {
+        input: {
+          ...input,
```
I recommend not using the spread operator here. Historically a lot bugs was caused by this. For example someone passes more parameters than necessary into input and Typescript does not complain or a developer enriches the input with other data that is needed for handling the mutation but is not part of the mutation variables.

alinhle commented on src/core/client/admin/routes/Stories/StoryActions/StoryActions.tsx:
```diff
@@ -56,7 +64,7 @@ const StoryActions: FunctionComponent<Props> = ({
                 </Localized>
               )}
-              {canClose && (
+              {canClose && !canUnarchive && (
```
Check for !canUnarchive seems unnecessary. canClose means Story is open, so canUnarchive is always false, because story would be closed.

alinhle commented on src/core/client/admin/routes/Stories/StoryActions/UnarchiveStoriesMutation.ts:
```diff
@@ -0,0 +1,54 @@
+      },
+      variables: {
+        input: {
+          ...input,
```
See comment on other mutation on this.

alinhle commented on src/core/client/stream/common/ModerateStream/ModerateStreamContainer.tsx:
```diff
@@ -46,6 +49,19 @@ const ModerateStreamContainer: FunctionComponent<Props> = ({
+        <Localized id="general-archived">
+          <span>Archived</span>
+        </Localized>
+      </Marker>
```
I think this is a good candidate to refector it out to outer scope.

e.g. 
```ts
const ArchiveMarker = () => (
<Marker color="warning" variant="filled">
        <Icon size="sm" className={styles.icon}>
          archive
        </Icon>
        <Localized id="general-archived">
          <span>Archived</span>
        </Localized>
      </Marker>
);

const ModerateStreamContainer: FunctionComponent<Props> = ({
[...]
  if (isArchived || isArchiving) {
    return <ArchiveMarker />;
  }
```

alinhle commented on src/core/client/stream/tabs/Comments/Comment/ArchivedReportFlow/ArchivedReportFlowContainer.tsx:
```diff
@@ -0,0 +1,111 @@
+            {settings.organization.url}
+          </a>
+        }
+        $orgURL={settings.organization.url}
```
<img width="616" alt="grafik" src="https://user-images.githubusercontent.com/14221600/134251081-57f07e15-9595-4411-bbd3-7c7e193f3bd7.png">

Should that render the organisation email or at least the organisation name instead? Rendering the organisation url here, and then linking to a mail address is very confusing.

alinhle commented on src/core/client/stream/tabs/Comments/Comment/CommentContainer.tsx:
```diff
@@ -708,14 +722,17 @@ export const CommentContainer: FunctionComponent<Props> = ({
             onClose={toggleShowReportFlow}
           />
         )}
+        {showReportFlow && story.isArchived && !story.isArchiving && (
```
Shouldn't this say `showReportFlow && (story.isArchived || story.isArchiving)`? Genuinely asking.

alinhle commented on src/core/client/stream/tabs/Comments/Comment/PermalinkButton/PermalinkCopyButton.tsx:
```diff
@@ -0,0 +1,84 @@
+  const [copied, setCopied] = useState(false);
+
+  const handleCopy = useCallback(() => {
+    setCopied(true);
```
I think we should add a timeout here that restores the `non copied` state for the user to click again if needed. 

The timeout should be set to around 800 to avoid issues with the permalink popover auto close timeout.

alinhle commented on src/core/client/stream/tabs/Comments/Comment/PermalinkButton/PermalinkCopyButton.tsx:
```diff
@@ -0,0 +1,84 @@
+    <CopyToClipboard text={permalinkURL} onCopy={handleCopy}>
+      <Button
+        className={cn(
+          CLASSES.sharePopover.copyButton,
```
Props should accept `className` and `CLASSES.sharePopover.copyButton` should be passed by the Popover.

alinhle commented on src/core/client/stream/tabs/Comments/Comment/PermalinkButton/PermalinkPopover.tsx:
```diff
@@ -46,56 +43,24 @@ const PermalinkPopover: FunctionComponent<Props> = ({
-      <CopyToClipboard text={permalinkURL} onCopy={handleCopy}>
-        <Button
-          className={cn(
-            CLASSES.sharePopover.copyButton,
```
This className should be passed into the PermalinkcopyButton.

alinhle commented on src/core/client/stream/tabs/Comments/Comment/ArchivedReportFlow/ArchivedReportFlowContainer.tsx:
```diff
@@ -0,0 +1,111 @@
+        <div className={styles.heading}>Need a link to this comment?</div>
+      </Localized>
+
+      <PermalinkCopyButton
```
add `className={CLASSES.reportPopover.copyLinkButton}`

alinhle commented on src/core/client/stream/tabs/Configure/ArchivedConfigurationContainer.tsx:
```diff
@@ -0,0 +1,81 @@
+          <Localized id="configure-archived-title">
+            This comment stream has been archived
+          </Localized>
+        }
```
You could make this accessible by:
```suggestion
        color="warning"
        aria-labelledby="configure-archived-title"
        title={
          <Localized id="configure-archived-title">
            <div id="configure-archived-title">
              This comment stream has been archived
            </div>
          </Localized>
        }
```

alinhle commented on src/core/client/stream/tabs/Configure/Configure.tsx:
```diff
@@ -21,21 +22,43 @@ export interface Props {
+      </div>
+    );
+  }
+
```
This is just picky, but could be more dry if you want:

```ts
const UserBoxElement = <UserBoxContainer viewer={viewer} settings={settings} />;
```

and then use `{UserBoxElement}` for both occasions.

alinhle commented on src/core/client/stream/tabs/Configure/OpenOrCloseStream/OpenOrCloseStreamContainer.tsx:
```diff
@@ -100,6 +104,8 @@ const enhanced = withFragmentContainer<Props>({
       id
       isClosed
+      isArchiving
+      isArchived
```
Seems like changes in this file is no longer relevant, as we don't show this container anymore?

alinhle commented on src/core/client/stream/tabs/Configure/UnarchiveStoriesMutation.ts:
```diff
@@ -0,0 +1,54 @@
+      },
+      variables: {
+        input: {
+          ...input,
```
This spread is a comment source of bugs, I recommend explicitly specifying the input properties.

alinhle commented on src/core/server/app/handlers/api/comment/featured.ts:
```diff
@@ -61,20 +61,20 @@ export const featuredHander = ({
 
     const comments = await retrieveFeaturedComments(
-      mongo,
+      mongo.live,
```
So featured comments are no longer available once the story has been archived?

nick-funk replied:
This is a REST handler for the dashboard. It does not use archive because the dashboard does not use archived data.

alinhle commented on src/core/server/app/middleware/passport/index.ts:
```diff
@@ -38,7 +38,7 @@ export function createPassport(options: Options): passport.Authenticator {
 
   // Use the LocalStrategy.
-  auth.use(createLocalStrategy(options));
+  auth.use(createLocalStrategy({ ...options, mongo: options.mongo.live }));
```
Looks like all the other authors accept `MongoContext` instead, maybe we should adapt `createLocalStrategy` to behave similiarly as well?

alinhle commented on src/core/server/app/handlers/api/story/ratings.ts:
```diff
@@ -91,7 +91,7 @@ export const ratingsJSONPHandler = ({
     }
 
-    const ratings = await retrieveStoryRatings(mongo, tenant.id, story.id);
+    const ratings = await retrieveStoryRatings(mongo.live, tenant.id, story.id);
```
I think anything related to ratings will break when the story is archieved.

nick-funk replied:
That is not the case and I have tested it and it works.

alinhle replied:
Actually it is, the ratings become zero:
<img width="1400" alt="Screenshot 2021-10-05 at 21 44 33" src="https://user-images.githubusercontent.com/14221600/136095844-8d5efc37-6f10-43f0-8963-ac1c4d01e3f3.png">

nick-funk replied:
Check now, should be fixed after recent changes.

alinhle commented on src/core/server/app/handlers/api/user/download.ts:
```diff
@@ -42,7 +42,7 @@ export const userDownloadHandler = ({
 
       // Send the export down the response.
-      await sendUserDownload(res, mongo, tenant, user, latestContentDate);
+      await sendUserDownload(res, mongo.live, tenant, user, latestContentDate);
```
Download GDPR data is broken as it does not include archived comments.

alinhle commented on src/core/server/services/users/delete.ts:
```diff
@@ -232,19 +253,27 @@ export async function deleteUser(
-  await deleteUserActionCounts(mongo, userID, tenantID);
+  await deleteUserActionCounts(mongo.live, userID, tenantID);
+  if (mongo.archive) {
+    await deleteUserActionCounts(mongo.archive, userID, tenantID);
```
This should fail, as it would target the wrong collection.

alinhle commented on src/core/server/services/users/delete.ts:
```diff
@@ -262,5 +291,15 @@ export async function deleteUser(
+  // Delete the user's archived comments.
+  if (mongo.archive) {
+    await deleteUserComments(mongo, redis, userID, tenantID, now, true);
+  }
```
Looks like duplicate?

alinhle commented on src/core/server/services/users/users.ts:
```diff
@@ -1640,7 +1641,7 @@ export async function updateUserLastCommentID(
  * @param user the User that we're looking up the limit for
  */
 export async function retrieveUserLastComment(
```
Mhh maybe call this `retrieveUserLastCommentNotArchived`, long name but clear...

alinhle commented on src/core/server/stacks/createComment.ts:
```diff
@@ -187,17 +188,21 @@ export default async function create(
   }
 
+  if (story.isArchiving || story.isArchived) {
+    throw new Error("Cannot create comments on an archive story");
```
Would be great, to make this a standardized error in  `src/core/server/errors/index.ts`

alinhle commented on src/core/server/stacks/editComment.ts:
```diff
@@ -134,14 +134,18 @@ export default async function edit(
   }
 
+  if (story.isArchiving || story.isArchived) {
+    throw new Error("Cannot create comments on an archive story");
```
Would be great, to make this a standardized error in  `src/core/server/errors/index.ts`

alinhle commented on src/core/server/cron/autoArchiving.ts:
```diff
@@ -0,0 +1,115 @@
+  const amount = config.get("auto_archiving_batch_size");
+  const age = config.get("auto_archive_older_than");
+
+  // For each of the tenant's, process their users notifications.
```
Looks like a c&p comment

alinhle commented on src/core/server/graph/loaders/Comments.ts:
```diff
@@ -190,21 +192,34 @@ export default (ctx: GraphContext) => ({
+    if (storyID) {
+      story = await ctx.loaders.Stories.story.load(storyID);
+    }
+
```
Maybe we should return an empty connection here when `story.isArchiving` is true? Because the data will be inconsistent.

alinhle replied:
The frontend could also show a nice ui saying `apologies, this story is currently being archived`.

nick-funk replied:
It now returns an empty connection

alinhle commented on src/core/server/graph/loaders/Comments.ts:
```diff
@@ -243,20 +258,33 @@ export default (ctx: GraphContext) => ({
+    if (!story) {
+      throw new Error("cannot get comments for a story that doesn't exist");
+    }
+
```
Same here and the other occassions on this file

alinhle commented on src/core/server/models/comment/comment.ts:
```diff
@@ -389,24 +389,28 @@ export async function editComment(
+  }
 
-  return ids.map((id) => comments.find((comment) => comment.id === id) || null);
+  return [];
```
Why was this change made? I see logic that depends on the nulls in the array in other parts of the codebase.

alinhle commented on src/core/server/models/comment/counts/counts.ts:
```diff
@@ -226,3 +227,52 @@ export async function updateRelatedCommentCounts<
+
+  if (commentCounts.action) {
+    for (const key in commentCounts.action) {
+      if (Object.prototype.hasOwnProperty.call(commentCounts.action, key)) {
```
Early return: I would do `if (!Object.prototype.hasOwnProperty.call(commentCounts.action, key)) { continue; })`. But this is fine too.

alinhle commented on src/core/server/models/comment/counts/counts.ts:
```diff
@@ -226,3 +227,52 @@ export async function updateRelatedCommentCounts<
   return result.value || null;
 }
+
+export const negateCommentCounts = (options: {
```
I wonder if lodash has something for doing the negate easier.

alinhle commented on src/core/server/models/story/helpers.ts:
```diff
@@ -67,6 +67,20 @@ export function getStoryClosedAt(
+    return false;
+  }
+
+  return story.isArchiving ?? false;
```
Intersting , thanks for showing me the`??`. This is essentially the same as `Boolean(story.isArchiving)` in this case right?

nick-funk replied:
Correct, it is the same in this case.

alinhle commented on src/core/server/models/story/story.ts:
```diff
@@ -727,3 +735,124 @@ export async function retrieveStorySections(
+      $set: {
+        isArchiving: false,
+        isArchived: true,
+        archivedAt: now,
```
I wonder if we should set 

```ts
{
  unarchivedAt: null,
  startedUnarchivingAt: null,
}
```

nick-funk replied:
It would seem like that's a good idea at first glance. However, we use the unarchivedAt as a filter to prevent auto-archiving from running on stories that admin's specifically decide they want to unarchive. Otherwise, whenever they unarchive a story, the auto-archiver would just archive it again in an hour or two.

alinhle commented on src/core/server/models/story/story.ts:
```diff
@@ -727,3 +735,124 @@ export async function retrieveStorySections(
+      $set: {
+        isArchiving: false,
+        isArchived: false,
+        unarchivedAt: now,
```
I wonder if we should set 

```ts
{
  archivedAt: null,
  startedArchivingAt: null,
}
```

nick-funk replied:
I'm not sure I like removing this metadata. It would prevent me from seeing when a story was archived and when it is un-archived. @wyattjoh largely asked me to add these in to aid in recover-ability and debugging should archiving issues crop up. Removing this data doesn't make the document more confusing in my mind. The reason being that if we have an `archivedAt` date that is less than the `unarchivedAt` date, we have all the data we need to grab both the archiving and un-archiving logs from Sentry/Cloud-Console. This seems like a good thing to keep.

alinhle commented on src/core/server/services/comments/comments.ts:
```diff
@@ -79,3 +110,174 @@ export async function removeTag(
+    const archivedComment = await retrieveCommentModel(
+      mongo.archive,
+      tenantID,
+      id
```
This should not work, as it would access the wrong collection.

alinhle commented on src/core/server/services/comments/comments.ts:
```diff
@@ -79,3 +110,174 @@ export async function removeTag(
+    comment,
+    paginationParameters
+  );
+}
```
I kinda like to see comments here for the new methods introduced.

alinhle commented on src/core/server/services/mongodb/collections.ts:
```diff
@@ -18,16 +18,26 @@ export const tenants = createCollection<Tenant>("tenants");
 
 export const comments = createCollection<Comment>("comments");
 
+export const archivedComments = createCollection<Comment>("archivedComments");
+
 export const stories = createCollection<Story>("stories");
 
 export const commentActions = createCollection<CommentAction>("commentActions");
 
+export const archivedCommentActions = createCollection<CommentAction>(
+  "archivedCommentActions"
+);
+
 export const sites = createCollection<Site>("sites");
 
 export const commentModerationActions = createCollection<
   CommentModerationAction
 >("commentModerationActions");
 
+export const archivedCommentModerationActions = createCollection<
+  CommentModerationAction
+>("archivedCommentModerationActions");
+
```
It seems passing mongo around in the code now with two instances is still very awkward and error prone.

Here is a crazy idea: 
I would throughout the whole application only pass `MongoContext` instead of `Db`, everywhere. 

For this I would adapt `createCollection` that would accept also `MongoContext` instead of `Db` and then select the appropriate database automatically.

So the api would look like:
 ```ts
export const commentModerationActions = createCollection<
  CommentModerationAction
>("commentModerationActions");

export const archivedCommentModerationActions = createCollection<
  CommentModerationAction
>("archivedCommentModerationActions", { archive: true });
```

Now we essentially tie the collection to the specific db instance, which also reflects reality, and we don't ever have to care about knowing which mongo db instance to choose from throughout the entire codebase.

alinhle commented on src/core/client/admin/routes/Stories/StoryActions/ArchiveStoriesMutation.ts:
```diff
@@ -0,0 +1,54 @@
+            stories {
+              id
+              status
+              isArchived
+              isArchiving
```
`isArchived` and `isArchiving` should be also added to the `CloseStoryMutation` and `OpenStoryMutation` so that in case they fail because of archiving, the status is immediately updated.

alinhle reviewed (changes requested):
This touched quite a bit of stuff, still, looking good so far. 

Before fixing other things, you might want to look at this comment here: https://github.com/coralproject/talk/pull/3666#discussion_r713517312
This could make things hell lot of less confusing and easier.

Some things I have noted while doing the review.
- Rating is probably broken
- Reactions are gone for achived story?
- Feature comments are gone for achived story?

alinhle commented on src/core/client/stream/tabs/Configure/OpenOrCloseStream/OpenOrCloseStreamContainer.tsx:
```diff
@@ -48,6 +48,10 @@ const OpenOrCloseStreamContainer: FunctionComponent<Props> = ({
 
+  if (story.isArchiving || story.isArchived) {
+    return null;
+  }
```
This check seems now to be obsolete as this Component is longer rendered when story isArchived or in progress of archiving. If you wanna a check here, I'd recommend you to render an error instead.

nick-funk commented:
> This touched quite a bit of stuff, still, looking good so far.
> 
> Before fixing other things, you might want to look at this comment here: [#3666 (comment)](https://github.com/coralproject/talk/pull/3666#discussion_r713517312)
> This could make things hell lot of less confusing and easier.
> 
> Some things I have noted while doing the review.
> 
>     * Rating is probably broken
> 
>     * Reactions are gone for achived story?
> 
>     * Feature comments are gone for achived story?

Addressing some of your comments here:

"Rating is probably broken"
Try setting the `ENABLE_RATINGS_AND_REVIEWS` feature flag, R&R appears to be working fine on this branch. ~~I think the submission flow might be bugged though, so I'll have to fix that up and make it aware of archiving state.~~ Did some testing and archiving seems to work with ratings and reviews.

"Reactions are gone for achived story?"
Correct, you cannot react to an archived story.

"Feature comments are gone for achived story?"
~~Hmmm, these do seem to be missing. I'll take a quick peek, should be able to get this fixed.~~
This has been fixed, featured comments now show.

nick-funk committed:
Merge remote-tracking branch 'origin/develop' into spike/archiving

nick-funk committed:
Compute tag counts correctly for archived stories

nick-funk committed:
Fix archived deletion of user action counts

nick-funk committed:
Specify mongoLive for moderateComments action

nick-funk committed:
Set storyIDs explicitly in archive mutations

nick-funk committed:
Refactor out archived marker into independent function

nick-funk committed:
Use org name for archived reporting flow email link

nick-funk committed:
Show archived report flow when story is archiving

nick-funk committed:
Reset the permalink copy botton after 800 ms

nick-funk committed:
Make archived configure callout more accessible

nick-funk committed:
Remove un-necessary archiving checks

The OpenOrCloseStreamContainer is never shown when a
story is archived.

nick-funk committed:
Update createLocalStrategy to use MongoContext

nick-funk committed:
Rename retrieveUserLastComment to retrieveUserLastCommentNotArchived

nick-funk committed:
Throw proper error when creating comment on archived story

We don't allow creating comments on archived stories, so when
this happen, throw a proper error with appropriate translation
keys.

nick-funk committed:
Remove errant comment

nick-funk committed:
Fix models retrieveComment to use correct collection

nick-funk committed:
Preliminarily bring in collections to MongoContext

nick-funk committed:
Replace null-able returns for retreiveManyComments

nick-funk committed:
Make user download work with archived comments

alinhle commented on src/core/client/stream/tabs/Configure/ArchivedConfigurationContainer.tsx:
```diff
@@ -0,0 +1,84 @@
+        icon={
+          <Icon size="sm" className={styles.icon}>
+            archive
+          </Icon>
```
Should add `container="section"`

alinhle commented on src/core/server/services/users/delete.ts:
```diff
@@ -36,25 +37,29 @@ interface Batch {
+    }
   }
 
   const cursor = collections
```
I think the collection needs to be switched here for archived

nick-funk replied:
Yeah, I fixed this up.

alinhle commented on src/core/server/services/users/delete.ts:
```diff
@@ -36,25 +37,29 @@ interface Batch {
-      batch.stories
-    );
-    batch.stories = [];
+    if (!isArchived) {
```
Why don't we run this for `isArchived`?

nick-funk replied:
We don't process story count/state modifications on archived stories. We also don't apply `moderate(...)` to approve/reject comments that are deleted when archived cause it will screw up the mod queue counts. Andrew and Wyatt agreed that if a story is archived, it's done and no one is likely to go back and re-review a story that is older than the archiving limit.

alinhle commented on src/core/server/services/comments/comments.ts:
```diff
@@ -79,3 +109,175 @@ export async function removeTag(
+    tenantID,
+    ids
+  );
+  if (liveComments.length > 0) {
```
Should filter out null elements, before this check. (I know it's confusing as hell)

alinhle replied:
Sth like `liveComments.filter(c => Boolean(c)).length > 0`

nick-funk replied:
Yeah, found that when testing the permalink view, it's fixed up now.

nick-funk committed:
Convert all usages of collection to use context collections

nick-funk committed:
Fix bug for archived permalink view

Because of us reintroducing nulls in the return array
for retrieveManyCommentModels, we have to check that in
the services/comments/comments/retrieveManyComments(...)

nick-funk committed:
Bring back archiving state on close/open story mutations

nick-funk committed:
Return empty comments connection during archiving step

nick-funk committed:
Add accessibility section container to archived callout

nick-funk committed:
Fix ratings retrieval for archived stories

nick-funk committed:
Replace custom errors with StoryNotFoundError

nick-funk committed:
storyID -> comment.storyID

No idea how this compiled and ran

nick-funk committed:
Last trick to get R&R working properly with archiving

nick-funk committed:
Merge remote-tracking branch 'origin/develop' into spike/archiving

# Conflicts:
#	src/core/server/app/authenticators/facebook/facebook.ts
#	src/core/server/app/authenticators/google/google.ts
#	src/core/server/app/handlers/api/auth/local/link.ts
#	src/core/server/app/middleware/passport/strategies/verifiers/oidc.ts
#	src/core/server/app/middleware/passport/strategies/verifiers/sso.ts
#	src/core/server/config.ts
#	src/core/server/services/oidc/oidc.ts
#	src/core/server/services/users/users.ts

nick-funk committed:
Fix mutations, update rejected queue snapshot

nick-funk committed:
Allow passing classname to PermalinkCopyButton

Also use this to pass in the copyButton CSS styling
class name.

nick-funk committed:
Remove duplication of UserBoxContainer

alinhle commented on src/core/client/stream/tabs/Comments/Comment/ArchivedReportFlow/ArchivedReportFlowContainer.tsx:
```diff
@@ -84,6 +85,7 @@ const ArchivedReportFlowContainer: FunctionComponent<Props> = ({
         variant="outlined"
         paddingSize="extraSmall"
         upperCase
+        className={CLASSES.sharePopover.copyButton}
```
This one needs its own CLASSES name.

alinhle replied:
Like `CLASSES.reportPopover.copyLinkButton`

alinhle commented:
`Open Story` should not show, when Story is archived.

<img width="233" alt="Screenshot 2021-10-05 at 21 37 25" src="https://user-images.githubusercontent.com/14221600/136233921-e6c8945c-527e-4758-b436-4ede61735e47.png">

Resolved

nick-funk committed:
Set CSS classname for archive report flow copy button

nick-funk committed:
Simplify the archive flow for admin > stories popover

Also fix bug where you could open a story if it was
archived.

nick-funk committed:
Throw an error if you try to open an archived story

nick-funk committed:
Remove outdated snapshot

alinhle committed:
fix: in archive make reactions featured and answered readonly

alinhle committed:
chore: remove collections

alinhle committed:
chore: rename copy button class

nick-funk committed:
Allow some moderation actions to occur when archived

alinhle committed:
Merge branch 'spike/archiving' of https://github.com/coralproject/talk into spike/archiving

nick-funk committed:
Disable mod card decision section for archived comments

nick-funk committed:
Show archived callout in approved/rejected queues

When filtering by story in mod queues for an archived
story, let the user know they are seeing an archived
story.

nick-funk committed:
Add comments to methods in services/comments

nick-funk committed:
Remove redundant canArchive check

alinhle committed:
fix: Decision history should work when comment revisions get deleted

nick-funk committed:
Make archived marker show for all users

nick-funk committed:
Fix comment doc linter warning over missing newline

nick-funk committed:
Prevent users merging archived stories

alinhle committed:
fix: add translations to new errors

alinhle committed:
fix: delete moderation actions when removing story

nick-funk committed:
Show archived callout for all mode queues

No longer shows comments for approve/rejected queue.

nick-funk committed:
Increase default auto-archiving interval/batch size

- Interval by default runs 4 times an hour
- Attempts to archive 500 stories every interval

alinhle committed:
fix: snapshots

alinhle committed:
fix: consider lastCommentedAt before createdAt for auto archiving

nick-funk committed:
Stop lock collisions when auto-archiving stories

alinhle commented on src/core/server/models/comment/comment.ts:
```diff
@@ -939,9 +966,49 @@ export async function removeCommentTag(
+    );
+  }
+
+  return [...liveCounts, ...archivedCounts];
```
Counts must be returned back in original order, or counts will be switched.

nick-funk replied:
This should be properly ordered and sorted now :D

alinhle commented on src/core/server/queue/tasks/rejector.ts:
```diff
@@ -31,16 +28,22 @@ export interface RejectorData {
+      after: connection ? connection.pageInfo.endCursor : undefined,
+    },
+    false
+  );
```
Banning users should reject their archived comments as well. We can reuse the method we used for `deleteUser`. (Or can be done later)

nick-funk replied:
this is fixed up now :)

nick-funk committed:
Preserve storyID order when retrieving comment tag counts

nick-funk committed:
Make the rejector work with archived comments

alinhle committed:
fix: delete user counts

nick-funk committed:
Use regular rejectComment in rejector for live comments

nick-funk committed:
Update all comment counts in rejector

nick-funk committed:
Remove unused logger in rejector helper function

alinhle committed:
chore: tiny refactors
